autonomi/src/networking/driver/task_handler.rs

```diff
@@ -133,11 +129,12 @@ impl TaskHandler {
         Ok(())
     }
 
+    /// Returns true if the task with QueryId is finished
     pub fn update_get_record(
         &mut self,
         id: QueryId,
         res: Result<kad::GetRecordOk, kad::GetRecordError>,
-    ) -> Result<(), TaskHandlerError> {
+    ) -> Result<bool, TaskHandlerError> {
         match res {
```

Comment:
Returning a bool to indicate we need to kad cleanup in the layer above here has the same effect but without ANY complexity.